compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -760,6 +760,15 @@ object Checking {
     // enum extending a value class type (AnyVal or an alias of it)
     // The error message 'EnumMayNotBeValueClassesID' will take care of generating the error message (See #22236)
     if (clazz.isDerivedValueClass && !clazz.isEnumAnonymClass) {
+      val parentOpt = cdef.rhs match {
+        case impl: Template =>
+          impl.parents.headOption
+        case _ => None
+      }
+      val isExtendingAliasOfAnyVal = parentOpt.exists { parent =>
+        parent.symbol.isAliasType && parent.tpe.nn.dealias =:= defn.AnyValType
+      }
```

Comment:
To avoid the `Option` (incurred for every VC), you could `inline def checkSyntax(): Unit = rhs match etc error()` and then use it below. 

I haven't checked if we only need to check the first parent. I see that `clazz.info` does not help, and now I understand the remark that `AnyVal` is just syntax for inline class. The sibling tickets are for incorrectly not creating a companion (here) or creating one unnecessarily (the other ticket).

It's too bad to add this extra check or restriction. Oh, I wonder if the check really is that the value class has a companion. That is the bug? Right, it compiles if user adds a companion.

Reply:
`clazz` will always have a companion because of constructor proxies, but it only has a real companion if
```
clazz.companionModule.info.decls.exists(_.isConstructor)
```
So that might suffice as a check.

Speculatively, typer could add the companion tree at that point. I guess extmethods needs a tree to see the module, even though it is all symbolic fiction.